File: java_solutions_1/529.minesweeper.java
/*
 * @lc app=leetcode id=529 lang=java
 *
 * [529] Minesweeper
 */

class Solution {
    public char[][] updateBoard(char[][] board, int[] click) {
        return java.util.stream.Stream.of(0).map(_v -> new java.util.concurrent.atomic.AtomicReference<java.util.function.Consumer<int[]>>()).peek(ref -> ref.set(pos -> java.util.stream.Stream.of(0).filter(_x -> pos[0] >= 0 && pos[0] < board.length && pos[1] >= 0 && pos[1] < board[0].length).peek(_x -> java.util.stream.Stream.of(0).filter(_y -> board[pos[0]][pos[1]] == 'M').forEach(_y -> board[pos[0]][pos[1]] = 'X')).peek(_x -> java.util.stream.Stream.of(0).filter(_y -> board[pos[0]][pos[1]] == 'E').map(_y -> java.util.stream.IntStream.rangeClosed(-1, 1).flatMap(r -> java.util.stream.IntStream.rangeClosed(-1, 1).map(c -> pos[0] + r >= 0 && pos[0] + r < board.length && pos[1] + c >= 0 && pos[1] + c < board[0].length && board[pos[0] + r][pos[1] + c] == 'M' ? 1 : 0)).sum()).peek(mines -> board[pos[0]][pos[1]] = mines > 0 ? (char) (mines + '0') : 'B').filter(mines -> mines == 0).forEach(_y -> java.util.stream.IntStream.rangeClosed(-1, 1).forEach(r -> java.util.stream.IntStream.rangeClosed(-1, 1).forEach(c -> ref.get().accept(new int[]{pos[0] + r, pos[1] + c}))))).forEach(_x -> {}))).peek(ref -> ref.get().accept(click)).map(ref -> board).findFirst().get();
    }
}

File: kotlin_solutions/529.minesweeper.kt
/*
 * @lc app=leetcode id=529 lang=kotlin
 *
 * [529] Minesweeper
 */

class Solution { fun updateBoard(board: Array<CharArray>, click: IntArray) = click.let { (r, c) -> DeepRecursiveFunction<Pair<Int, Int>, Unit> { (i, j) -> if (i in board.indices && j in board[0].indices && board[i][j] == 'E') { listOf(-1 to -1, -1 to 0, -1 to 1, 0 to -1, 0 to 1, 1 to -1, 1 to 0, 1 to 1).count { (di, dj) -> (i + di) in board.indices && (j + dj) in board[0].indices && board[i + di][j + dj] == 'M' }.let { mines -> if (mines == 0) { 'B'.also { board[i][j] = it }.also { listOf(-1 to -1, -1 to 0, -1 to 1, 0 to -1, 0 to 1, 1 to -1, 1 to 0, 1 to 1).forEach { (di, dj) -> callRecursive(i + di to j + dj) } } } else board[i][j] = '0' + mines } } }(r to c).run { if (board[r][c] == 'M') board[r][c] = 'X' } .run{board} } }
